packages/flutter/lib/src/foundation/diagnostics.dart

```diff
@@ -2108,10 +2028,72 @@ abstract class TreeDiagnosticsMixin implements TreeDiagnostics {
   /// }
   /// ```
   ///
-  /// Used by [toStringDeep], [toDiagnosticsNode] and [toStringShallow].
+  /// Used by [toDiagnosticsNode] and [toString].
   @protected
   @mustCallSuper
   void debugFillProperties(List<DiagnosticsNode> properties) { }
+}
+
+/// A class that can be used as a mixin that helps dump string and
+/// [DiagnosticsNode] representations of trees.
+///
+/// Use this class as a mixin in any context where it is possible to use mixins.
+abstract class DiagnosticableTree extends Diagnosticable {
+  /// Abstract const constructor. This constructor enables subclasses to provide
+  /// const constructors so that they can be used in const expressions.
+  const DiagnosticableTree();
+
+  /// Returns a one-line detailed description of the object.
+  ///
+  /// This description is often somewhat long. This includes the same
+  /// information given by [toStringDeep], but does not recurse to any children.
+  ///
+  /// The [toStringShallow] method can take an argument, which is the string to
+  /// place between each part obtained from [debugFillProperties]. Passing a
+  /// string such as `'\n '` will result in a multiline string that indents the
+  /// properties of the object below its name (as per [toString]).
+  ///
+  /// See also:
+  ///
+  ///  * [toString], for a brief description of the object.
+  ///  * [toStringDeep], for a description of the subtree rooted at this object.
+  String toStringShallow([String joiner = '; ']) {
+    final StringBuffer result = new StringBuffer();
+    result.write(toString());
+    result.write(joiner);
+    final List<DiagnosticsNode> properties = <DiagnosticsNode>[];
+    debugFillProperties(properties);
+    result.write(
+      properties.where((DiagnosticsNode n) => !n.hidden).join(joiner),
+    );
+    return result.toString();
+  }
+
+  /// Returns a string representation of this node and its descendants.
+  ///
+  /// The [toStringDeep] method takes arguments, but those are intended for
+  /// internal use when recursing to the descendants, and so can be ignored.
+  ///
+  /// See also:
+  ///
+  ///  * [toString], for a brief description of the object but not its children.
+  ///  * [toStringShallow], for a detailed description of the object but not its
+  ///    children.
+  String toStringDeep([String prefixLineOne = '', String prefixOtherLines = '']) {
+    return toDiagnosticsNode().toStringDeep(prefixLineOne, prefixOtherLines);
+  }
+
+  @override
+  String toShortDescription() => describeIdentity(this);
```

Comment:
Done. Please look at the test output diff to make sure the output is reasonable. I kept this change as its own commit on my branch so it is easy to view in isolation.
I'm now hiding the key property from Widget as it is redundant. New output looks generally reasonable.